stetho/src/main/java/com/facebook/stetho/inspector/domstorage/DOMStoragePeerManager.java

```diff
@@ -0,0 +1,166 @@
+      // tags that will be observed by the peer.  We can fix this by making the page frames
+      // dynamically update in response to DOMStorage events.  This would also allow us to
+      // add new SharedPreferences tags as we observe them being created by way of
+      // android.os.FileObserver.
+      List<String> tags = SharedPreferencesHelper.getSharedPreferenceTags(mContext);
+      for (String tag : tags) {
+        SharedPreferences prefs = mContext.getSharedPreferences(tag, Context.MODE_PRIVATE);
+        DevToolsSharedPreferencesListener listener =
+            new DevToolsSharedPreferencesListener(prefs, tag);
+        prefs.registerOnSharedPreferenceChangeListener(listener);
+        mPrefsListeners.add(listener);
+      }
+    }
+
+    @Override
+    protected synchronized void onLastPeerUnregistered() {
+      for (DevToolsSharedPreferencesListener prefsListener : mPrefsListeners) {
+        prefsListener.unregister();
+      }
+      mPrefsListeners.clear();
+    }
+  };
+
+  private class DevToolsSharedPreferencesListener
+      implements SharedPreferences.OnSharedPreferenceChangeListener {
+    private final SharedPreferences mPrefs;
+    private final DOMStorage.StorageId mStorageId;
+
+    /**
+     * Maintains a copy of the prefs data structure so that we can invoke
+     * {@code DOMStorage.domStorageItemUpdated}.  This method requires that we know the old
+     * value to perform updates.  Using {@code domStorageItemRemoved}/{@code Added} causes a UI
+     * glitch where the item is moved to the end of the list, unfortunately.
+     */
+    private final Map<String, Object> mCopy;
+
+    public DevToolsSharedPreferencesListener(SharedPreferences prefs, String tag) {
+      mPrefs = prefs;
+      mStorageId = new DOMStorage.StorageId();
+      mStorageId.securityOrigin = tag;
+      mStorageId.isLocalStorage = true;
+      mCopy = prefsCopy(prefs.getAll());
+    }
+
+    public void unregister() {
+      mPrefs.unregisterOnSharedPreferenceChangeListener(this);
+    }
+
+    @Override
+    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
+      Map<String, ?> entries = sharedPreferences.getAll();
+      boolean existedBefore = mCopy.containsKey(key);
+      boolean existsNow = entries.containsKey(key);
+      Object newValue = existsNow ? entries.get(key) : null;
+      if (existedBefore && existsNow) {
+        signalItemUpdated(
+            mStorageId,
+            key,
+            SharedPreferencesHelper.valueToString(mCopy.get(key)),
+            SharedPreferencesHelper.valueToString(newValue));
+        mCopy.put(key, newValue);
+      } else if (existedBefore) {
+        signalItemRemoved(mStorageId, key);
+        mCopy.remove(key);
+      } else if (existsNow) {
+        signalItemAdded(
+            mStorageId,
+            key,
+            SharedPreferencesHelper.valueToString(newValue));
+        mCopy.put(key, newValue);
+      } else {
+        CLog.writeToConsole(
```

Comment:
This should be a throw, or a statement to file a bug. If both `existedBefore` and `existsNow` are `false` that has to imply Android did something it shouldn't do, or more likely our map is out of date.